File: android/app/src/main/java/com/entrenadordigital/app/NativeBridge.kt
package com.entrenadordigital.app

import android.content.Context
import android.os.Build
import android.os.VibrationEffect
import android.os.Vibrator
import android.webkit.JavascriptInterface

class NativeBridge(
    private val activity: MainActivity,
    private val backupManager: BackupManager
) {
    @JavascriptInterface
    fun setTrainingMode(enabled: Boolean) {
        activity.runOnUiThread { activity.setTrainingMode(enabled) }
    }

    @JavascriptInterface
    fun vibrate(milliseconds: Int) {
        if (milliseconds <= 0) return
        val duration = milliseconds.coerceAtMost(1000).toLong()
        val vibrator = activity.getSystemService(Context.VIBRATOR_SERVICE) as Vibrator
        @Suppress("DEPRECATION")
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            vibrator.vibrate(VibrationEffect.createOneShot(duration, VibrationEffect.DEFAULT_AMPLITUDE))
        } else {
            vibrator.vibrate(duration)
        }
    }

    @JavascriptInterface
    fun getAppVersion(): String = BuildConfig.VERSION_NAME

    @JavascriptInterface
    fun saveBackup(json: String, suggestedName: String) {
        activity.runOnUiThread { backupManager.beginSave(json, suggestedName) }
    }

    @JavascriptInterface
    fun openBackup() {
        activity.runOnUiThread { backupManager.beginOpen() }
    }

    @JavascriptInterface
    fun finishApp() {
        activity.runOnUiThread { activity.finish() }
    }
}

File: web/src/platform/native-bridge.ts
export interface AndroidBridge {
  setTrainingMode?: (enabled: boolean) => void;
  vibrate?: (milliseconds: number) => void;
  getAppVersion?: () => string;
  checkForUpdates?: () => void;
  saveBackup?: (json: string, suggestedName: string) => void;
  openBackup?: () => void;
  finishApp?: () => void;
}

declare global {
  interface Window {
    Android?: AndroidBridge;
  }
}

export function nativeBridge(): AndroidBridge | undefined {
  return window.Android;
}

export function setNativeTrainingMode(enabled: boolean): void {
  try {
    nativeBridge()?.setTrainingMode?.(enabled);
  } catch {
    // Native integration is optional in browser preview.
  }
}

export function nativeVibrate(milliseconds: number): boolean {
  try {
    const bridge = nativeBridge();
    if (!bridge?.vibrate) return false;
    bridge.vibrate(milliseconds);
    return true;
  } catch {
    return false;
  }
}

export function getNativeAppVersion(): string | null {
  try {
    return nativeBridge()?.getAppVersion?.() ?? null;
  } catch {
    return null;
  }
}

export function requestNativeUpdateCheck(): boolean {
  try {
    const bridge = nativeBridge();
    if (!bridge?.checkForUpdates) return false;
    bridge.checkForUpdates();
    return true;
  } catch {
    return false;
  }
}
